lib/kube/proxy/pod_filters.go

```diff
@@ -0,0 +1,306 @@
+	deniedResources []types.KubernetesResource
+	// log is the logger.
+	log logrus.FieldLogger
+}
+
+// FilterBuffer receives a byte array, decodes the response into the appropriate
+// type and filters the resources based on allowed and denied rules configured.
+// After filtering them, it serializes the response and dumps it into output buffer.
+// If any error occurs, the call returns an error.
+func (d *podFilterer) FilterBuffer(buf []byte, output io.Writer) error {
+	// decode the response into the appropriate Kubernetes API type.
+	obj, bf, err := d.decode(buf)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// if bf is not empty, it means that response does not contain any valid response
+	// and it should be safe to write it back into the buffer.
+	if len(bf) > 0 {
+		_, err = output.Write(buf)
+		return trace.Wrap(err)
+	}
+
+	var filtered runtime.Object
+	switch o := obj.(type) {
+	case *metav1.Status:
+		// Status object is returned when the Kubernetes API returns an error and
+		// should be forwarded to the user.
+		filtered = obj
+	case *corev1.Pod:
+		// filterPod filters a single corev1.Pod and returns an error if access to
+		// pod was denied.
+		if err := filterPod(o, d.allowedResources, d.deniedResources); err != nil {
+			if !trace.IsAccessDenied(err) {
+				d.log.WithError(err).Warnf("Unable to compile role kubernetes_resources.")
+			}
+			return trace.Wrap(err)
+		}
+	case *corev1.PodList:
+		filtered = filterCoreV1PodList(o, d.allowedResources, d.deniedResources, d.log)
+	case *metav1.Table:
+		filtered, err = d.filterMetav1Table(o, d.allowedResources, d.deniedResources)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	default:
+		return trace.BadParameter("unexpected type received; got %T", obj)
+	}
+	// encode the filterer response back to the user.
+	return d.encode(filtered, output)
+}
+
+// FilterObj receives a runtime.Object type and filters the resources on it
+// based on allowed and denied rules resources.
+// After filtering them, the obj is manipulated to hold the filtered information.
+func (d *podFilterer) FilterObj(obj runtime.Object) (bool, error) {
+	switch o := obj.(type) {
+	case *corev1.Pod:
+		err := filterPod(o, d.allowedResources, d.deniedResources)
+		if err != nil && !trace.IsAccessDenied(err) {
+			d.log.WithError(err).Warnf("Unable to compile role kubernetes_resources.")
+		}
+		// if err is not nil we should not include it.
+		return err == nil, nil
+	case *corev1.PodList:
+		_ = filterCoreV1PodList(o, d.allowedResources, d.deniedResources, d.log)
+		return len(o.Items) > 0, nil
+	case *metav1.Table:
+		_, err := d.filterMetav1Table(o, d.allowedResources, d.deniedResources)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+		return len(o.Rows) > 0, nil
+	default:
+		return false, trace.BadParameter("unexpected type received; got %T", obj)
+	}
+}
+
+// decode decodes the buffer into the appropriate type if the responseCode is valid.
+// If not valid, returns the buffer unchanged.
```

Comment:
What indicates "not valid"? Is it an error or a nil runtime.Object without an error?